### PR TITLE
Link unfurling nodejs sample

### PR DESCRIPTION
**Pull request by Prithvi-MSFT**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -178,6 +178,7 @@ The [Teams Toolkit](https://marketplace.visualstudio.com/items?itemName=TeamsDev
 |22| Activity feed broadcast |Demonstrating a feature to notify any message to all members of the organisation using activity feed notification .|[View][graph-activity-feed-broadcast#csharp]|[View][graph-activity-feed-broadcast#js]|
 |23|App complete auth|This sample demos authentication feature in bot,tab and messaging extension.|[View][app-complete-auth#cs]|
 |24|App identity linking with sso|This sample demos mapping with aad id, facebook, and google account of user in bot, ME and tab.||[View][app-identity-link-with-sso#nodejs]|
+|25|Link unfurling share to teams|This sample demos the feature of link unfurling in share to teams.||[View][link-unfurling-stt#nodejs]|
```
Comment (v-chetsh) [inline README.md:181]: This is confusing, can you add proper name like Link **unfurling for share to teams**

File: samples/link-unfurling-stt/nodejs/README.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+- nodejs
+- javascript
+extensions:
+contentType: samples
+createdDate: "14-03-2022 00:15:13"
+---
+
+# Link unfruling Share to teams sample
+
+This sample demos the feature of link unfurling in share to teams.
```
Comment (v-chetsh) [inline samples/link-unfurling-stt/nodejs/README.md:18]: Explain a bit more on what exactly is "share to teams"

File: samples/link-unfurling-stt/nodejs/README.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    ```bash
+    npm install
+    ```
+
+4) Run ngrok - point to port 3978
+
+    ```bash
+    ngrok http -host-header=rewrite 3978
+    ```
+5) Update the `.env` configuration for the bot to use the `MicrosoftAppId` and `MicrosoftAppPassword`, `BaseUrl` with application base url. For e.g., your ngrok url. (Note the MicrosoftAppId is the AppId created in step 1 (Setup for Bot), the MicrosoftAppPassword is referred to as the "client secret" in step 1 (Setup for Bot) and you can always create a new client secret anytime.). Also update `AppId` and `EntityId` where Appid will be the manifest id and entity id will be the entity id of tab (Here entity id is tab).
```
Comment (v-chetsh) [inline samples/link-unfurling-stt/nodejs/README.md:59]: Why do we need these 2 values?
These are not present in C# sample

File: samples/link-unfurling-stt/nodejs/appPackage/manifest.json [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  "developer": {
+    "name": "Microsoft",
+    "websiteUrl": "https://example.azurewebsites.net",
+    "privacyUrl": "https://example.azurewebsites.net/privacy",
+    "termsOfUseUrl": "https://example.azurewebsites.net/termsofuse"
+  },
+  "icons": {
+    "color": "icon-color.png",
+    "outline": "icon-outline.png"
+  },
```
Comment (v-chetsh) [inline samples/link-unfurling-stt/nodejs/appPackage/manifest.json:16]: Your actual icons are with different name "color.png"/"outline.png"
Kindly update it properly.

File: samples/link-unfurling-stt/nodejs/bots/teamsBot.js [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+                    await context.sendActivity("Hello and welcome! This sample demonstrates the use of link unfurling in share to teams scope.");
+                }
+            }
+
+            await next();
+        });
+    }
+
+    async handleTeamsAppBasedLinkQuery(context, query) {
+        var userCard = CardFactory.adaptiveCard(this.getLinkUnfurlingCard());
```
Comment (v-chetsh) [inline samples/link-unfurling-stt/nodejs/bots/teamsBot.js:22]: Add summary for when it invokes.

File: samples/link-unfurling-stt/nodejs/bots/teamsBot.js [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+            {
+              "type": "Image",
+              "url": `${process.env.ApplicationBaseUrl}/Images/report.png`
+            }
+          ],
+          "actions":[
+            {
+                "type": "Action.OpenUrl",
+                "title": "Open tab",
+                "url": `https://teams.microsoft.com/l/entity/${process.env.AppId}/${process.env.EntityId}?webUrl=${process.env.ApplicationBaseUrl}/AuthTab`
```
Comment (v-chetsh) [inline samples/link-unfurling-stt/nodejs/bots/teamsBot.js:58]: How app Id differs from App client id?

File: samples/link-unfurling-stt/nodejs/views/tab.ejs [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        crossorigin="anonymous"></script>
+    <script async defer src="https://teams.microsoft.com/share/launcher.js"></script>
+    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.min.js"></script>
+    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/4.5.2/js/bootstrap.min.js"></script>
+    <script>
+
+        $(document).ready(function () {
+            microsoftTeams.initialize();
+        });
+
```
Comment (v-chetsh) [inline samples/link-unfurling-stt/nodejs/views/tab.ejs:15]: Remove extra line space here.

File: samples/link-unfurling-stt/nodejs/views/tab.ejs [hunk 8]
```diff
@@ -0,0 +1,39 @@
+<head>
+    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.5.2/css/bootstrap.min.css">
+    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">
+    <script src="https://statics.teams.cdn.office.net/sdk/v1.10.0/js/MicrosoftTeams.min.js"
+        integrity="sha384-6oUzHUqESdbT3hNPDDZUa/OunUj5SoxuMXNek1Dwe6AmChzqc6EJhjVrJ93DY/Bv"
+        crossorigin="anonymous"></script>
+    <script async defer src="https://teams.microsoft.com/share/launcher.js"></script>
+    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.min.js"></script>
+    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/4.5.2/js/bootstrap.min.js"></script>
+    <script>
```
Comment (v-chetsh) [inline samples/link-unfurling-stt/nodejs/views/tab.ejs:10]: Verify if you are using all the CDN links.

**Review by v-chetsh — COMMENTED** [review]
Added my comments

File: samples/link-unfurling-stt/nodejs/views/tab.ejs [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+        }
+    </style>
+</head>
+
+<body class="theme-light">
+    <div class="surface">
+        <h3 id="tabheader"> Link unfurling in STT</h3>
+        <img id="reportimg" src="/Images/report.png" />
+        <div id="stt-btn" class="teams-share-button" data-href="https://6073-116-75-10-155.ngrok.io?krs" data-button-type="medium" data-icon-px-size="160" data-preview="true"></div>
+    </div>
```
[the remark this answers is not in the document]
Comment (NikitaJain-MSFT) [reply samples/link-unfurling-stt/nodejs/views/tab.ejs:38]: updated url with botframework domain which is added in manifest

File: README.md [hunk 10]
```diff
@@ -178,6 +178,7 @@ The [Teams Toolkit](https://marketplace.visualstudio.com/items?itemName=TeamsDev
 |22| Activity feed broadcast |Demonstrating a feature to notify any message to all members of the organisation using activity feed notification .|[View][graph-activity-feed-broadcast#csharp]|[View][graph-activity-feed-broadcast#js]|
 |23|App complete auth|This sample demos authentication feature in bot,tab and messaging extension.|[View][app-complete-auth#cs]|
 |24|App identity linking with sso|This sample demos mapping with aad id, facebook, and google account of user in bot, ME and tab.||[View][app-identity-link-with-sso#nodejs]|
+|25|Link unfurling for 'Share to teams'|This sample demos the feature of link unfurling for Share to teams.||[View][link-unfurling-stt#nodejs]|
```
Comment (Wajeed-msft) [inline README.md:181]: Please take care of renaming similar to C# solution.

https://github.com/OfficeDev/Microsoft-Teams-Samples/pull/273#discussion_r834960825
https://github.com/OfficeDev/Microsoft-Teams-Samples/pull/273#discussion_r834960423

File: samples/link-unfurling-stt/nodejs/bots/teamsBot.js [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+        this.onMembersAdded(async (context, next) => {
+            const membersAdded = context.activity.membersAdded;
+            for (let member = 0; member < membersAdded.length; member++) {
+                if (membersAdded[member].id !== context.activity.recipient.id) {
+                    await context.sendActivity("Hello and welcome! This sample demonstrates the use of link unfurling in share to teams scope.");
+                }
+            }
+
+            await next();
+        });
```
Comment (Wajeed-msft) [inline samples/link-unfurling-stt/nodejs/bots/teamsBot.js:18]: We don't have Bot mentioned in manifest, could you please remove code which is not being used?

File: samples/link-unfurling-stt/nodejs/appPackage/manifest.json [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+          "parameters": [
+            {
+              "name": "searchQuery",
+              "title": "Search",
+              "description": "Your search query",
+              "inputType": "text"
+            }
+          ]
+        }
+      ],
```
Comment (Wajeed-msft) [inline samples/link-unfurling-stt/nodejs/appPackage/manifest.json:60]: Can we remove this?

File: samples/link-unfurling-stt/nodejs/appPackage/manifest.json [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+    "short": "Link unfurling STT"
+  },
+  "description": {
+    "short": "Sample to test link unfurling in STT",
+    "full": "This sample demos the feature of link unfurling in share to teams"
+  },
+  "accentColor": "#FFFFFF",
+  "staticTabs": [
+    {
+      "contentUrl": "https://<<base-url>>/tab",
```
Comment (Wajeed-msft) [inline samples/link-unfurling-stt/nodejs/appPackage/manifest.json:27]: We are already providing https in contentURL , should we change <<base-url> to <<domain-name>> to keep it consistent?

File: samples/link-unfurling-stt/nodejs/bots/teamsBot.js [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+        var card = {
+          "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
+          "type": "AdaptiveCard",
+          "version": "1.0",
+          "body": [
+            {
+              "type": "TextBlock",
+              "size": "Medium",
+              "weight": "Bolder",
+              "text": "The analytics details are"
```
Comment (Wajeed-msft) [inline samples/link-unfurling-stt/nodejs/bots/teamsBot.js:49]:
```suggestion
              "text": "Analytics details:"
```

File: samples/link-unfurling-stt/nodejs/bots/teamsBot.js [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+            }
+        }
+    }
+
+    // Adaptive card for link unfurling.
+    getLinkUnfurlingCard = () => {
+        var card = {
+          "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
+          "type": "AdaptiveCard",
+          "version": "1.0",
```
Comment (Wajeed-msft) [inline samples/link-unfurling-stt/nodejs/bots/teamsBot.js:43]: Could you please use the latest version?

File: samples/link-unfurling-stt/nodejs/views/tab.ejs [hunk 16]
```diff
@@ -0,0 +1,33 @@
+<head>
+    <script src="https://statics.teams.cdn.office.net/sdk/v1.10.0/js/MicrosoftTeams.min.js"
+        integrity="sha384-6oUzHUqESdbT3hNPDDZUa/OunUj5SoxuMXNek1Dwe6AmChzqc6EJhjVrJ93DY/Bv"
+        crossorigin="anonymous"></script>
```
Comment (Wajeed-msft) [inline samples/link-unfurling-stt/nodejs/views/tab.ejs:4]: Check if need this, if not please remove it.

File: samples/link-unfurling-stt/nodejs/views/tab.ejs [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+		
+        #reportimg {
+            margin-left: 28rem;
+        }
+		
+        #stt-btn {
+            margin-left: 40rem;
+            margin-top: 2rem;
+        }
+    </style>
```
Comment (Wajeed-msft) [inline samples/link-unfurling-stt/nodejs/views/tab.ejs:24]: Please move the styles to different .css file.

File: samples/link-unfurling-stt/nodejs/views/tab.ejs [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+</head>
+
+<body class="theme-light">
+    <div class="surface">
+        <h3 id="tabheader"> Link unfurling in STT</h3>
+        <img id="reportimg" src="/Images/report.png" width="500" height="425"/>
+        <div id="stt-btn" class="teams-share-button" data-href="https://profile.botframework.com?hello" data-button-type="medium" data-icon-px-size="160" data-preview="true"></div>
```
Comment (Wajeed-msft) [inline samples/link-unfurling-stt/nodejs/views/tab.ejs:31]: Can we remove the reference to Bot Framework completely from solution and use the ngrok URL (ApplicationBaseUrl from .env file)?

**Review by Wajeed-msft — CHANGES_REQUESTED** [review]

File: samples/app-link-unfurling-in-share-to-teams/nodejs/bots/teamsBot.js [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+              "url": `${process.env.ApplicationBaseUrl}/Images/report.png`
+            }
+          ],
+          "actions":[
+            {
+                "type": "Action.OpenUrl",
+                "title": "Open tab",
+                "url": `https://teams.microsoft.com/l/entity/${process.env.MicrosoftAppId}/tab?webUrl=${process.env.ApplicationBaseUrl}/AuthTab`
+              }
+          ]
```
Comment (ht4963-ms) [inline samples/app-link-unfurling-in-share-to-teams/nodejs/bots/teamsBot.js:52]: Can we include another action to open this in a stage view? (Sample copied from doc https://docs.microsoft.com/en-us/microsoftteams/platform/tabs/tabs-link-unfurling#:~:text=through%20deep%20link-,Invoke%20Stage%20View%20from%20Adaptive%20Card,the%20stage%20as%20a%20tab.)
`
    type: "Action.Submit",
    name: "View",
    data: {
          msteams: {
            type: "invoke",
            value: {
                type: "tab/tabInfoAction",
                tabInfo: {
                    contentUrl: contentUrl,
                    websiteUrl: websiteUrl,
                    name: "Tasks",
                    entityId: "entityId"
                 }
                }
            }
        }
} `

File: samples/app-link-unfurling-in-share-to-teams/nodejs/README.md [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+
+6) Run your app
+
+    ```bash
+    npm start
+    ```
+- **Manually update the manifest.json**
+    - Edit the `manifest.json` contained in the  `appPackage/` folder to replace with your MicrosoftAppId (that was created in step1.1 and is the same value of MicrosoftAppId in `.env` file) *everywhere* you see the place holder string `{MicrosoftAppId}` (depending on the scenario the Microsoft App Id may occur multiple times in the `manifest.json`)
+    - Zip up the contents of the `appPackage/` folder to create a `manifest.zip`
+    - Upload the `manifest.zip` to Teams (in the left-bottom *Apps* view, click "Upload a custom app")
```
Comment (ht4963-ms) [inline samples/app-link-unfurling-in-share-to-teams/nodejs/README.md:69]: I think this instruction is slightly outdated

File: samples/app-link-unfurling-in-share-to-teams/nodejs/bots/teamsBot.js (hunk 19 above)
Comment (Prithvi-MSFT) [reply samples/app-link-unfurling-in-share-to-teams/nodejs/bots/teamsBot.js:70]: Added button to open tab in stage view

File: samples/app-link-unfurling-in-share-to-teams/nodejs/README.md (hunk 20 above)
Comment (Prithvi-MSFT) [reply samples/app-link-unfurling-in-share-to-teams/nodejs/README.md]: changed

**Comment by v-chetsh** [comment]
@ht4963-ms - Can you please review and approve this PR?

**Comment by v-chetsh** [comment]
@ht4963-ms - Can you please review and approve this PR?

**Comment by Prithvi-MSFT** [comment]
**Build status** - All builds passed
https://github.com/OfficeDev/Microsoft-Teams-Samples/actions/runs/4200665528

File: samples/app-link-unfurling-in-share-to-teams/nodejs/README.md [hunk 21]
```diff
@@ -0,0 +1,106 @@
...
+- office-365
+languages:
+- nodejs
+- javascript
+extensions:
+contentType: samples
+createdDate: "14-12-2022 00:15:13"
+---
+
+# Link unfruling Share to teams sample
```
Comment (ChetanSharma-msft) [inline samples/app-link-unfurling-in-share-to-teams/nodejs/README.md:16]: `## Link unfurling Share to teams sample`

File: samples/app-link-unfurling-in-share-to-teams/nodejs/appPackage/manifest.json [hunk 22]
```diff
@@ -0,0 +1,65 @@
+{
+  "$schema": "https://developer.microsoft.com/json-schemas/teams/v1.15/MicrosoftTeams.schema.json",
+  "manifestVersion": "1.15",
+  "version": "1.0.0",
+  "id": "<<Microsoft-App-Id>>",
+  "packageName": "com.microsoft.teams.linkunfurlingstt",
+  "developer": {
+    "name": "Microsoft",
+    "websiteUrl": "https://example.azurewebsites.net",
```
Comment (ChetanSharma-msft) [inline samples/app-link-unfurling-in-share-to-teams/nodejs/appPackage/manifest.json:9]: Update URL's to use Microsoft links

File: samples/app-link-unfurling-in-share-to-teams/nodejs/assets/sample.json [hunk 23]
```diff
@@ -0,0 +1,72 @@
+[
+  {
+    "name": "officedev-microsoft-teams-samples-app-link-unfurling-in-share-to-teams-nodejs",
```
Comment (ChetanSharma-msft) [inline samples/app-link-unfurling-in-share-to-teams/nodejs/assets/sample.json:3]: This is not exist in readme file.

File: samples/app-link-unfurling-in-share-to-teams/nodejs/assets/sample.json [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+    "source": "officeDev",
+    "title": "App link unfurling in share to teams",
+    "shortDescription": "Demonstrating feature of link unfurling in share to teams.",
+    "url": "https://github.com/OfficeDev/Microsoft-Teams-Samples/tree/main/samples/app-link-unfurling-in-share-to-teams/nodejs",
+    "longDescription": [
+      "Demonstrating feature of link unfurling in share to teams."
```
Comment (ChetanSharma-msft) [inline samples/app-link-unfurling-in-share-to-teams/nodejs/assets/sample.json:9]: "This sample shows the feature of link unfurling in share to teams."

File: samples/app-link-unfurling-in-share-to-teams/nodejs/assets/sample.json [hunk 25]
```diff
@@ -0,0 +1,72 @@
...
+        "key": "TEAMS-FEATURES",
+        "value": "tab"
+      }
+    ],
+    "thumbnails": [
+      {
+        "type": "image",
+        "order": 100,
+        "url": "https://raw.githubusercontent.com/OfficeDev/Microsoft-Teams-Samples/main/samples/app-link-unfurling-in-share-to-teams/nodejs/Images/app-link-unfurling-stt.gif",
+        "alt": "Solution UX showing check in information of user"
```
Comment (ChetanSharma-msft) [inline samples/app-link-unfurling-in-share-to-teams/nodejs/assets/sample.json:43]: This content is wrong.

File: samples/app-link-unfurling-in-share-to-teams/nodejs/bots/teamsBot.js [hunk 26]
```diff
@@ -0,0 +1,77 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (ChetanSharma-msft) [inline samples/app-link-unfurling-in-share-to-teams/nodejs/bots/teamsBot.js:1]: Add proper header

**Review by ChetanSharma-msft — CHANGES_REQUESTED** [review]
Added my review comments on your PR.

File: samples/app-link-unfurling-in-share-to-teams/nodejs/bots/teamsBot.js (hunk 26 above)
Comment (Prithvi-MSFT) [reply samples/app-link-unfurling-in-share-to-teams/nodejs/bots/teamsBot.js:1]: Added

File: samples/app-link-unfurling-in-share-to-teams/nodejs/assets/sample.json (hunk 25 above)
Comment (Prithvi-MSFT) [reply samples/app-link-unfurling-in-share-to-teams/nodejs/assets/sample.json:43]: Changed

File: samples/app-link-unfurling-in-share-to-teams/nodejs/assets/sample.json (hunk 24 above)
Comment (Prithvi-MSFT) [reply samples/app-link-unfurling-in-share-to-teams/nodejs/assets/sample.json:9]: Changed

File: samples/app-link-unfurling-in-share-to-teams/nodejs/appPackage/manifest.json (hunk 22 above)
Comment (Prithvi-MSFT) [reply samples/app-link-unfurling-in-share-to-teams/nodejs/appPackage/manifest.json:9]: Updated

File: samples/app-link-unfurling-in-share-to-teams/nodejs/README.md (hunk 21 above)
Comment (Prithvi-MSFT) [reply samples/app-link-unfurling-in-share-to-teams/nodejs/README.md:16]: Changed

**Review by ChetanSharma-msft — APPROVED** [review]
Changes looks fine, approving it.

**Comment by ChetanSharma-msft** [comment]
@Prithvi-MSFT - Please resolve conflicts and run build pipeline.

**Pull request merged**